pkg/resource/deploy/step_generator.go

```diff
@@ -1180,6 +1184,97 @@ func (sg *stepGenerator) getProviderResource(urn resource.URN, provider string)
 	return result
 }
 
+// initErrorSpecialKey is a special property key used to indicate that a diff is due to
+// intiialization errors existing in the old state instead of due to a specific proeprty
+// diff between old and new states.
+const initErrorSpecialKey = "#initerror"
+
+// applyReplaceOnChanges adjusts a DiffResult returned from a provider to apply the ReplaceOnChange
+// settings in the desired state and initerrors from the previous state.
+func (sg *stepGenerator) applyReplaceOnChanges(diff plugin.DiffResult,
+	replaceOnChanges []string, hasInitErrors bool) (plugin.DiffResult, error) {
+
+	var err error
+	diffPaths := map[string]resource.PropertyPath{}
+	for p := range diff.DetailedDiff {
+		diffPaths[p], err = resource.ParsePropertyPath(p)
+		if err != nil {
+			return diff, err
+		}
+	}
+
+	var replaceOnChangePaths []resource.PropertyPath
+	for _, p := range replaceOnChanges {
+		path, err := resource.ParsePropertyPath(p)
+		if err != nil {
+			return diff, err
+		}
+		replaceOnChangePaths = append(replaceOnChangePaths, path)
+	}
+
+	// Calculate the new DetailedDiff
+	modifiedDiff := map[string]plugin.PropertyDiff{}
+	for p, v := range diff.DetailedDiff {
+		diffPath, err := resource.ParsePropertyPath(p)
+		if err != nil {
+			return diff, err
+		}
+		for _, replaceOnChangePath := range replaceOnChangePaths {
+			if replaceOnChangePath.Contains(diffPath) {
+				v = v.ToReplace()
+			}
+			modifiedDiff[p] = v
+		}
+	}
+
+	// Calculate the new ReplaceKeys
+	modifiedReplaceKeysMap := map[resource.PropertyKey]struct{}{}
+	for _, k := range diff.ReplaceKeys {
+		modifiedReplaceKeysMap[k] = struct{}{}
+	}
+	for _, k := range diff.ChangedKeys {
+		for _, replaceOnChangePath := range replaceOnChangePaths {
+			keyPath, err := resource.ParsePropertyPath(string(k))
+			if err != nil {
+				continue
```

Comment:
Should we silently ignore parse errors?

Reply:
I'm not 100% sure what the requirements are on the `ChangedKeys` returned from a provider, and whether it might be legal for those to be names which cannot be parsed as paths.  If they do have a name that cannot be parsed as a path, it's not clear an error here is actionable, vs. just assuming that the name cannot be `replaceOnChange`d.  That said - it's possible that `ChangedKeys` are semantically only keys, not paths, and therefore that we should not be parsing these at all - and should just construct a `PropertyPath{string(k)}` by hand to represent this?  Curious for @pgavlin input on this (and the rest of the similar assumptions in this function).

Reply:
Yeah, `ChangedKeys` might contain invalid paths. IIRC, though, this is only used for rendering purposes, so being imprecise here is okay. The most important thing w.r.t. replace semantics is that there are either a) replaces present in the detailed diff, or b) entries in `RepalceKeys`.